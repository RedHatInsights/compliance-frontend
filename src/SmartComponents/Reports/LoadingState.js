import React from 'react';
import {
    PageHeader,
    PageHeaderTitle,
    Main
} from '@redhat-cloud-services/frontend-components';
import {
    Grid
} from '@patternfly/react-core';
import {
    LoadingComplianceCards
} from 'PresentationalComponents';

export const LoadingState = () => (
    <React.Fragment>
        <PageHeader>
            <PageHeaderTitle title="Compliance" />
        </PageHeader>
        <Main>
            <div className="policies-donuts">
                <Grid gutter='md'>
                    <LoadingComplianceCards/>
                </Grid>
            </div>
        </Main>
    </React.Fragment>
);

export default LoadingState;
