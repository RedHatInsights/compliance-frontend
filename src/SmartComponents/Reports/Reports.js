import React from 'react';
import {
    ReportTabs,
    LoadingComplianceCards,
    ReportCard,
    CompliancePoliciesEmptyState,
    StateViewWithError,
    StateViewPart
} from 'PresentationalComponents';
import { PageHeader, PageHeaderTitle, Main } from '@redhat-cloud-services/frontend-components';
import routerParams from '@redhat-cloud-services/frontend-components-utilities/files/RouterParams';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import {
    Grid,
    GridItem
} from '@patternfly/react-core';

const QUERY = gql`
{
    profiles(search: "has_test_results = true", limit: 1000){
        edges {
            node {
                id
                name
                refId
                description
                totalHostCount
                compliantHostCount
                majorOsVersion
                complianceThreshold
                businessObjective {
                    id
                    title
                }
                policy {
                    id
                    benchmark {
                        version
                    }
                }
                benchmark {
                    version
                }
            }
        }

    }
}
`;

export const Reports = () => {
    const { data, error, loading } = useQuery(QUERY, { fetchPolicy: 'cache-and-network' });
    let reportCards;
    let pageHeader;

    if (!loading && data) {
        const profiles = data.profiles.edges.map(profile => profile.node).filter((profile) => profile.totalHostCount > 0);

        if (profiles.length) {
            pageHeader = <PageHeader className='page-header-tabs'>
                <PageHeaderTitle title="Compliance reports" />
                <ReportTabs/>
            </PageHeader>;
            reportCards = profiles.map(
                (profile, i) =>
                    <GridItem sm={12} md={12} lg={6} xl={4} key={i}>
                        <ReportCard
                            key={i}
                            profile={profile}
                        />
                    </GridItem>
            );
        } else {
            pageHeader = <PageHeader style={{ paddingBottom: 22 }}><PageHeaderTitle title="Compliance" /></PageHeader>;
            reportCards = <CompliancePoliciesEmptyState title={'No policies are reporting'} />;
        }
    }

    return <StateViewWithError stateValues={ { error, data, loading } }>
        <StateViewPart stateKey='loading'>
            <PageHeader>
                <PageHeaderTitle title="Compliance" />
            </PageHeader>
            <Main>
                <div className="policies-donuts">
                    <Grid gutter='md'>
                        <LoadingComplianceCards/>
                    </Grid>
                </div>
            </Main>
        </StateViewPart>
        <StateViewPart stateKey='data'>
            { pageHeader }
            <Main>
                <div className="policies-donuts">
                    <Grid gutter='md'>
                        { reportCards }
                    </Grid>
                </div>
            </Main>
        </StateViewPart>
    </StateViewWithError>;
};

export default routerParams(Reports);
