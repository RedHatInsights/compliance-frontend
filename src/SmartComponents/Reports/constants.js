import { gql } from '@apollo/client';

export const QUERY = gql`
  query R_Profiles($filter: String!) {
    profiles(search: $filter, limit: 1000) {
      edges {
        node {
          id
          name
          refId
          description
          policyType
          totalHostCount
          testResultHostCount
          compliantHostCount
          unsupportedHostCount
          osMajorVersion
          complianceThreshold
          businessObjective {
            id
            title
          }
          policy {
            id
            name
          }
          benchmark {
            id
            version
          }
        }
      }
    }
  }
`;
