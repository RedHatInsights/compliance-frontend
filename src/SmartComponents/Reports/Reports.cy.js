import React from 'react';
import Reports from './Reports';
import { IntlProvider } from 'react-intl';

import { MemoryRouter } from 'react-router-dom';
import { ApolloProvider } from '@apollo/client';
import { ApolloClient, HttpLink, InMemoryCache } from 'apollo-boost';
import { Provider } from 'react-redux';
import { COMPLIANCE_API_ROOT } from '@/constants';
import { init } from 'Store';
import fixtures from '../../../cypress/fixtures/reports.json';

const client = new ApolloClient({
  link: new HttpLink({
    credentials: 'include',
    uri: COMPLIANCE_API_ROOT + '/graphql',
  }),
  cache: new InMemoryCache(),
});
const mountComponent = () => {
  cy.mount(
    <Provider store={init().getStore()}>
      <IntlProvider locale={navigator.language}>
        <MemoryRouter>
          <ApolloProvider client={client}>
            <Reports />
          </ApolloProvider>
        </MemoryRouter>
      </IntlProvider>
    </Provider>
  );
};

describe('Reports table tests', () => {
  beforeEach(() => {
    cy.intercept('*', {
      statusCode: 201,
      body: {
        data: fixtures,
      },
    });
    mountComponent();
  });
  describe('defaults', () => {
    it('The Pathways table renders', () => {
      cy.get('table').should('have.length', 1);
    });
  });

  it('expect to render emptystate', () => {
    cy.intercept('*', {
      statusCode: 201,
      body: {
        data: fixtures,
      },
    });
    cy.get('table').should('have.length', 1);
  });
});
