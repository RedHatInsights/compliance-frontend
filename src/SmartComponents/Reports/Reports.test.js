import toJson from 'enzyme-to-json';
import { useQuery } from '@apollo/react-hooks';

import { Reports } from './Reports.js';

jest.mock('@apollo/react-hooks');

describe('Reports', () => {
    it('expect to render without error', () => {
        useQuery.mockImplementation(() => ({
            data: {
                allProfiles: [{
                    id: '1',
                    refId: '121212',
                    name: 'profile1',
                    description: 'profile description',
                    totalHostCount: 1,
                    complianceThreshold: 1,
                    compliantHostCount: 1,
                    businessObjective: {
                        id: '1',
                        title: 'BO 1'
                    }
                }]
            }, error: false, loading: false }));

        const wrapper = shallow(
            <Reports />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('expect to render emptystate', () => {
        useQuery.mockImplementation(() => ({
            data: {
                allProfiles: []
            }, error: false, loading: false }));

        const wrapper = shallow(
            <Reports />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('expect to render an error', () => {
        const error = {
            networkError: { statusCode: 500 },
            error: 'Test Error loading'
        };
        useQuery.mockImplementation(() => ({ data: {}, error, loading: false }));
        const wrapper = shallow(
            <Reports />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('expect to render loading', () => {
        useQuery.mockImplementation(() => ({ data: {}, error: false, loading: true }));
        const wrapper = shallow(
            <Reports />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

});
