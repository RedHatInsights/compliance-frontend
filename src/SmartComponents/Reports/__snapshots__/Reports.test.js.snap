// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reports expect to render emptystate 1`] = `
<StateViewWithError
  stateValues={
    Object {
      "data": Object {
        "profiles": Object {
          "edges": Array [],
        },
      },
      "error": false,
      "loading": false,
    }
  }
>
  <StateViewPart
    stateKey="loading"
  >
    <PageHeader>
      <PageHeaderTitle
        title="Compliance"
      />
    </PageHeader>
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        >
          <LoadingComplianceCards />
        </Grid>
      </div>
    </Connect(Main)>
  </StateViewPart>
  <StateViewPart
    stateKey="data"
  >
    <PageHeader
      style={
        Object {
          "paddingBottom": 22,
        }
      }
    >
      <PageHeaderTitle
        title="Compliance"
      />
    </PageHeader>
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        >
          <CompliancePoliciesEmptyState
            title="No policies are reporting"
          />
        </Grid>
      </div>
    </Connect(Main)>
  </StateViewPart>
</StateViewWithError>
`;

exports[`Reports expect to render loading 1`] = `
<StateViewWithError
  stateValues={
    Object {
      "data": undefined,
      "error": false,
      "loading": true,
    }
  }
>
  <StateViewPart
    stateKey="loading"
  >
    <PageHeader>
      <PageHeaderTitle
        title="Compliance"
      />
    </PageHeader>
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        >
          <LoadingComplianceCards />
        </Grid>
      </div>
    </Connect(Main)>
  </StateViewPart>
  <StateViewPart
    stateKey="data"
  >
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        />
      </div>
    </Connect(Main)>
  </StateViewPart>
</StateViewWithError>
`;

exports[`Reports expect to render without error 1`] = `
<StateViewWithError
  stateValues={
    Object {
      "data": Object {
        "profiles": Object {
          "edges": Array [
            Object {
              "node": Object {
                "businessObjective": Object {
                  "id": "1",
                  "title": "BO 1",
                },
                "complianceThreshold": 1,
                "compliantHostCount": 1,
                "description": "profile description",
                "id": "1",
                "name": "profile1",
                "refId": "121212",
                "totalHostCount": 1,
              },
            },
          ],
        },
      },
      "error": false,
      "loading": false,
    }
  }
>
  <StateViewPart
    stateKey="loading"
  >
    <PageHeader>
      <PageHeaderTitle
        title="Compliance"
      />
    </PageHeader>
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        >
          <LoadingComplianceCards />
        </Grid>
      </div>
    </Connect(Main)>
  </StateViewPart>
  <StateViewPart
    stateKey="data"
  >
    <PageHeader
      className="page-header-tabs"
    >
      <PageHeaderTitle
        title="Compliance reports"
      />
      <withRouter(Connect(RouterParams)) />
    </PageHeader>
    <Connect(Main)>
      <div
        className="policies-donuts"
      >
        <Grid
          gutter="md"
        >
          <GridItem
            key="0"
            lg={6}
            md={12}
            sm={12}
            xl={4}
          >
            <ReportCard
              key="0"
              profile={
                Object {
                  "businessObjective": Object {
                    "id": "1",
                    "title": "BO 1",
                  },
                  "complianceThreshold": 1,
                  "compliantHostCount": 1,
                  "description": "profile description",
                  "id": "1",
                  "name": "profile1",
                  "refId": "121212",
                  "totalHostCount": 1,
                }
              }
            />
          </GridItem>
        </Grid>
      </div>
    </Connect(Main)>
  </StateViewPart>
</StateViewWithError>
`;
